### PR TITLE
Bug/1426 error response for empty id in put v2 entities eid attrs attrname

### PR DESCRIPTION
**Pull request by kzangeli**
Various fixes about empty URI PATH components.
Fixes issues #1426, #1487 and more.

### CODE REVIEW

File: src/lib/rest/RestService.cpp [hunk 1]
```diff
@@ -352,6 +352,90 @@ static void filterRelease(ParseData* parseDataP, RequestType request)
...
+  }
+
+  return true;
+}
+
+
+/* ****************************************************************************
+*
+* compErrorDetect - 
+*/
```
Comment (fgalan) [inline src/lib/rest/RestService.cpp]: A bit cryptic... maybe comments over each if branch in the logic would help to understand how it works (making future maintenance easier).

File: test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+02a. /v2/entities//
+===================
+HTTP/1.1 404 Not Found
+Content-Length: 95
+Content-Type: application/json
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Date: REGEX(.*)
+
+{
+    "description": "The requested entity has not been found. Check type and id",
```
Comment (fgalan) [inline test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test]: I understand this shoulc be: 

```
400 Bad Request
{ "error": "BadRequest", "description: "entity id length: 0, min length supported: 1" }
```

File: test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test [hunk 3]
```diff
@@ -0,0 +1,566 @@
...
+07a. /v2/types//
+================
+HTTP/1.1 404 Not Found
+Content-Length: 58
+Content-Type: application/json
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Date: REGEX(.*)
+
+{
+    "description": "Entity type not found",
```
Comment (fgalan) [inline test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test]: I understand this should be:

```
400 Bad Request
{ "error": "BadRequest", "description: "entity type length: 0, min length supported: 1" }
```

File: test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test [hunk 4]
```diff
@@ -0,0 +1,566 @@
...
+Date: REGEX(.*)
+
+{
+    "description": "service not found",
+    "error": "BadRequest"
+}
+
+
+06b. /v2/entities/E1/attrs/A1/value///
+======================================
```
Comment (fgalan) [inline test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test]: Is the case `/v2/entities/E1/attrs//value` included? If not, should we?

Comment (fgalan) [inline test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test]: It seems to: https://github.com/telefonicaid/fiware-orion/pull/2314/files#diff-4c2a67b23b1d7b3b081ad601f6db3664R42

NTC

File: test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test (hunk 2 above)
Comment (kzangeli) [inline test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test]: Some test must be missing somewhere.
Nothing to do with this PR, but, what the heck, I can try to find the error within the PR.

File: test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test (hunk 3 above)
Comment (kzangeli) [inline test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test]: Same here, a check of validity of type must be missing

File: test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test]: Good. I am pretty confident I have all the 'interesting' cases in the tests

File: src/lib/rest/RestService.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/rest/RestService.cpp]: Function rewritten in 9460e2c

File: test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test (hunk 2 above)
Comment (fgalan) [inline test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test]: Maybe the length-check exists, but only for the payload (i.e. in the NGSIv2 parsing routines).

File: test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test (hunk 3 above)
Comment (fgalan) [inline test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test]: Maybe the length-check exists, but only for the payload (i.e. in the NGSIv2 parsing routines).

File: test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test (hunk 2 above)
Comment (kzangeli) [inline test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test]: Fixed in d23d1bb

File: test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test (hunk 3 above)
Comment (kzangeli) [inline test/functionalTest/cases/1426_empty_components_in_uri_path/double_slash_at_end_of_path.test]: Fixed in d23d1bb

File: src/lib/common/errorMessages.h [hunk 5]
```diff
@@ -26,7 +26,10 @@
 * Author: Orion dev team
 */
 
-#define MORE_MATCHING_ENT "More than one matching entity. Please refine your query"
-#define INVAL_CHAR_URI    "invalid character in URI"
+#define MORE_MATCHING_ENT   "More than one matching entity. Please refine your query"
+#define INVAL_CHAR_URI      "invalid character in URI"
+#define EMPTY_ENTITY_ID     "entity id length: 0, min length supported: 1"
```
Comment (fgalan) [inline src/lib/common/errorMessages.h]: What about if min length change in the future? I'd suggest to use MIN_ID_LEN to compose these messages, e.g:

```
#define EMPTY_ENTITY_ID     "entity id length: 0, min length supported: " MIX_ID_LEN
```

(Actually I'm not sure if this is the right way of concatenating #define constants)

File: src/lib/rest/RestService.cpp [hunk 6]
```diff
@@ -352,6 +352,89 @@ static void filterRelease(ParseData* parseDataP, RequestType request)
...
+      }
+    }
+    else if ((components == 5) && (compV[3] == "attrs"))  // URL: /v2/entities/<entity-id>/attrs/<attr-name>
+    {
+      std::string entityId = compV[2];
+      std::string attrName = compV[4];
+
+      if (entityId == "")
+      {
+        details = "entity id length: 0, min length supported: 1";
```
Comment (fgalan) [inline src/lib/rest/RestService.cpp]: Now that you have the constant, let's use it in all places, i.e.:

```
details = EMPTY_ENTITY_ID;
```

(Similarly in other `details` set statements in this function)

File: src/lib/common/errorMessages.h (hunk 5 above)
Comment (kzangeli) [inline src/lib/common/errorMessages.h]: Fixed in 0609929

File: src/lib/rest/RestService.cpp (hunk 6 above)
Comment (kzangeli) [inline src/lib/rest/RestService.cpp]: Fixed in 0609929

File: src/lib/common/limits.h [hunk 7]
```diff
@@ -92,20 +96,23 @@
...
 *
 * CONSTANTS RESTINIT - 
 */ 
-#define   DEFAULT_CONNECTION_MEM  	64   
-#define   DEFAULT_MAX_CONNECTIONS   	128   
-#define   DEFAULT_MHD_THREAD_POOLSIZE   128 
+#define   DEFAULT_CONNECTION_MEM  	    64
+#define   DEFAULT_MAX_CONNECTIONS   	  128
+#define   DEFAULT_MHD_THREAD_POOLSIZE   128
+
```
Comment (kzangeli) [inline src/lib/common/limits.h]: I don't know why github presents these lines like this.
Actually I do. There are tabs in the lines ... :-(
Tabs are strongly forbidden. 
One more commit on its way ...

File: src/lib/common/limits.h [hunk 8]
```diff
@@ -155,21 +167,23 @@
...
- * MIN_ID_LEN - min length of id and type for entities, attributes and metadata
- */
+*
+* MIN_ID_LEN - min length of id and type for entities, attributes and metadata
+*/
 #ifndef MIN_ID_LEN
 #define MIN_ID_LEN 1
 #endif
-
+#define MIN_ID_LEN_STRING "1"
```
Comment (kzangeli) [inline src/lib/common/limits.h]: This line is not necessary anymore, as the STR()/STR2 macros solve the problem.

Comment (kzangeli) [inline src/lib/common/limits.h]: Fixed in 60b5914

File: src/lib/common/limits.h (hunk 7 above)
Comment (kzangeli) [inline src/lib/common/limits.h]: Fixed in 60b5914

**Comment by fgalan** [comment]
LGTM

**Pull request merged**